File: ios/KacGunOlduWidget/WidgetSnapshot.swift
import Foundation
import SwiftUI

/// The App Group both the app and this extension can read. The app writes the
/// snapshot here (AppDelegate, "kac_gun_oldu/widgets" channel).
let appGroup = "group.com.emalabs.kacgunoldu"
let snapshotKey = "widgets.snapshot"
/// Days marked with the widgets' "Bugün yaptım" button, waiting for the app.
let marksKey = "widgets.marks"

/// What the widgets draw, as the app last wrote it
/// (lib/services/home_widgets.dart).
///
/// It holds each card's last record and interval, never a day count: the
/// count is worked out here against the entry's date, so a widget is right
/// after midnight without the app being opened. The rules are the app's own
/// (`statsFor` in lib/domain/logic.dart) — keep the two in step.
struct WidgetSnapshot {
  let theme: ThemeColors
  /// Show the "Bugün yaptım" button (Ayarlar → Ana ekran widget'ı).
  var doneButton = true
  let tiers: [Tier: TierColors]
  let cards: [RawCard]

  struct RawCard {
    let id: String
    let name: String
    let icon: String
    let last: DateComponents?
    let typical: Int?
  }

  static func load() -> WidgetSnapshot? {
    guard
      let raw = UserDefaults(suiteName: appGroup)?.string(forKey: snapshotKey),
      let data = raw.data(using: .utf8),
      let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
    else { return nil }
    return parse(json)
  }

  static func parse(_ o: [String: Any]) -> WidgetSnapshot? {
    if let v = o["v"] as? Int, v > 1 { return nil }
    guard let t = o["theme"] as? [String: Any], let tierJson = o["tiers"] as? [String: Any] else {
      return nil
    }
    func color(_ d: [String: Any], _ key: String) -> Color { Color(argb: (d[key] as? NSNumber)?.uint32Value ?? 0xFF000000) }
    let theme = ThemeColors(
      surface: color(t, "surface"),
      onSurface: color(t, "onSurface"),
      muted: color(t, "muted"),
      primary: color(t, "primary"),
      onPrimary: color(t, "onPrimary")
    )
    var tiers: [Tier: TierColors] = [:]
    for tier in Tier.allCases {
      guard let c = tierJson[tier.rawValue] as? [String: Any] else { continue }
      tiers[tier] = TierColors(bg: color(c, "bg"), ink: color(c, "ink"), ring: color(c, "ring"), track: color(c, "track"))
    }
    let cards = (o["cards"] as? [[String: Any]] ?? []).compactMap { c -> RawCard? in
      guard let id = c["id"] as? String, !id.isEmpty else { return nil }
      let typical = (c["typical"] as? NSNumber)?.intValue
      return RawCard(
        id: id,
        name: c["name"] as? String ?? "",
        icon: c["icon"] as? String ?? "spark",
        last: (c["last"] as? String).flatMap(dateKey),
        typical: typical.flatMap { $0 > 0 ? $0 : nil }
      )
    }
    return WidgetSnapshot(theme: theme, doneButton: o["doneButton"] as? Bool ?? true, tiers: tiers, cards: cards)
  }

  /// A widget's "Bugün yaptım": the card is done today. The snapshot's
  /// `last` moves to today so the widget redraws at once; the mark waits
  /// under `marksKey` until the app records it (CardStore._takeWidgetMarks).
  static func markDone(_ cardId: String, on date: Date = Date()) {
    guard let defaults = UserDefaults(suiteName: appGroup) else { return }
    let c = Calendar.current.dateComponents([.year, .month, .day], from: date)
    let day = String(format: "%04d-%02d-%02d", c.year ?? 0, c.month ?? 0, c.day ?? 0)
    if let raw = defaults.string(forKey: snapshotKey),
      let data = raw.data(using: .utf8),
      var json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
      var cards = json["cards"] as? [[String: Any]]
    {
      for i in cards.indices where cards[i]["id"] as? String == cardId {
        let last = cards[i]["last"] as? String ?? ""
        if last < day { cards[i]["last"] = day }
      }
      json["cards"] = cards
      if let out = try? JSONSerialization.data(withJSONObject: json), let text = String(data: out, encoding: .utf8) {
        defaults.set(text, forKey: snapshotKey)
      }
    }
    var marks: [[String: String]] = []
    if let raw = defaults.string(forKey: marksKey), let data = raw.data(using: .utf8),
      let list = try? JSONSerialization.jsonObject(with: data) as? [[String: String]]
    {
      marks = list
    }
    marks.append(["id": cardId, "day": day])
    if let out = try? JSONSerialization.data(withJSONObject: marks), let text = String(data: out, encoding: .utf8) {
      defaults.set(text, forKey: marksKey)
    }
  }

  /// Overdue first, then everything else, each by how far along it is.
  func ordered(on date: Date) -> [WidgetCard] {
    let all = cards.map { WidgetCard($0, on: date) }
    let byRatio: (WidgetCard, WidgetCard) -> Bool = { $0.ratio > $1.ratio }
    return all.filter(\.isLate).sorted(by: byRatio) + all.filter { !$0.isLate }.sorted(by: byRatio)
  }

  func colors(_ card: WidgetCard) -> TierColors {
    tiers[card.tier] ?? TierColors(bg: theme.surface, ink: theme.onSurface, ring: theme.primary, track: theme.onSurface.opacity(0.1))
  }

  /// `YYYY-MM-DD` → calendar date.
  private static func dateKey(_ s: String) -> DateComponents? {
    let parts = s.split(separator: "-").compactMap { Int($0) }
    guard parts.count == 3 else { return nil }
    return DateComponents(year: parts[0], month: parts[1], day: parts[2])
  }
}

enum Tier: String, CaseIterable { case fresh, calm, soon, late }

struct TierColors {
  let bg: Color
  let ink: Color
  let ring: Color
  let track: Color
}

struct ThemeColors {
  let surface: Color
  let onSurface: Color
  let muted: Color
  let primary: Color
  let onPrimary: Color

  /// Kiremit, the default theme — before the app has ever written a snapshot.
  static let kiremit = ThemeColors(
    surface: Color(argb: 0xFFFBF5F0),
    onSurface: Color(argb: 0xFF241F1B),
    muted: Color(argb: 0xFF87786C),
    primary: Color(argb: 0xFFB8492A),
    onPrimary: Color(argb: 0xFFFFF3EA)
  )
}

/// One card on a given day, with everything the app's `statsFor` derives.
struct WidgetCard: Identifiable {
  let id: String
  let name: String
  let icon: String
  let hasRecord: Bool
  let days: Int
  let ratio: Double
  let isLate: Bool
  let tier: Tier
  /// Ring fill, 0.03–1.
  let progress: Double
  /// Days left in the usual interval; negative past it, nil while learning.
  let remaining: Int?

  init(_ raw: WidgetSnapshot.RawCard, on date: Date) {
    id = raw.id
    name = raw.name
    icon = raw.icon
    let calendar = Calendar.current
    if let last = raw.last, let lastDate = calendar.date(from: last) {
      hasRecord = true
      let d = calendar.dateComponents([.day], from: calendar.startOfDay(for: lastDate), to: calendar.startOfDay(for: date)).day ?? 0
      days = max(0, d)
    } else {
      hasRecord = false
      days = 0
    }
    let typical = raw.typical
    if let typical, typical != 0 {
      ratio = Double(days) / Double(typical)
    } else {
      ratio = min(0.95, Double(days) / 30)
    }
    // `+1` keeps short-interval cards from flagging on a single late day.
    isLate = typical != nil && hasRecord && Double(days) > Double(typical ?? 0) * 1.25 + 1
    tier = isLate ? .late : ratio > 0.95 ? .soon : ratio < 0.5 ? .fresh : .calm
    progress = Double(min(100, max(3, Int((ratio * 100).rounded())))) / 100
    remaining = typical == nil || !hasRecord ? nil : typical! - days
  }

  /// The card's one status line — `CardStatus` in lib/widgets/card_status.dart.
  var status: String {
    guard hasRecord else { return "Henüz işaretlenmedi" }
    guard let r = remaining else { return "Ritim öğreniliyor" }
    if r > 0 { return "\(r) gün kaldı" }
    if r == 0 { return "Bugün sırası" }
    return "\(-r) gün geçti"
  }

  /// Due today or past due: drawn in the accent, bold.
  var urgent: Bool { (remaining ?? 1) <= 0 }

  /// Already recorded today: the "Bugün yaptım" button shows it's done.
  var doneToday: Bool { hasRecord && days == 0 }

  var dayText: String { hasRecord ? "\(days)" : "—" }
  var unitText: String { hasRecord ? "gün oldu" : "kayıt yok" }

  var link: URL {
    let id = id.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? id
    return URL(string: "kacgunoldu://app/card/\(id)")!
  }

  /// `CardStatus.color`: the accent when urgent (the ink itself on the solid
  /// overdue card), otherwise the ink, quietened.
  func statusColor(_ c: TierColors) -> Color {
    guard urgent else { return c.ink.opacity(0.6) }
    return tier == .late ? c.ink : c.ring
  }
}

extension Color {
  init(argb: UInt32) {
    self.init(
      .sRGB,
      red: Double((argb >> 16) & 0xFF) / 255,
      green: Double((argb >> 8) & 0xFF) / 255,
      blue: Double(argb & 0xFF) / 255,
      opacity: Double((argb >> 24) & 0xFF) / 255
    )
  }
}

File: android/app/src/main/kotlin/com/emalabs/kacgunoldu/widget/WidgetSnapshot.kt
package com.emalabs.kacgunoldu.widget

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.time.LocalDate
import java.time.temporal.ChronoUnit
import kotlin.math.min
import kotlin.math.roundToInt

/**
 * What the home screen widgets draw, as the app last wrote it
 * (lib/services/home_widgets.dart → MainActivity → here).
 *
 * The snapshot holds each card's last record and interval, never a day
 * count: the count is worked out here, against today, so a widget is right
 * after midnight without the app being opened. The rules are the app's own
 * (`statsFor` in lib/domain/logic.dart) — keep the two in step.
 */
internal class WidgetSnapshot(
    val theme: ThemeColors,
    /** Show the "Bugün yaptım" button (Ayarlar → Ana ekran widget'ı). */
    val doneButton: Boolean,
    /** The app's wording, in the language it is set to. */
    val strings: WidgetStrings,
    private val tiers: Map<Tier, TierColors>,
    cards: List<WidgetCard>,
) {
    /** Overdue first, then everything else, each by how far along it is. */
    val ordered: List<WidgetCard> = cards.filter { it.isLate }.sortedByDescending { it.ratio } +
        cards.filter { !it.isLate }.sortedByDescending { it.ratio }

    val lateCount: Int get() = ordered.count { it.isLate }

    fun colors(card: WidgetCard): TierColors = tiers[card.tier] ?: FALLBACK_TIER

    fun byId(id: String?): WidgetCard? = id?.let { wanted -> ordered.firstOrNull { it.id == wanted } }

    companion object {
        private const val PREFS = "kac_gun_oldu_widgets"
        private const val KEY = "snapshot"
        private const val MARKS = "marks"
        private const val VERSION = 1

        fun save(context: Context, json: String) {
            context.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit().putString(KEY, json).apply()
        }

        /**
         * A widget's "Bugün yaptım": the card is done today. The widget
         * redraws from the snapshot at once (its `last` moved to today); the
         * mark waits in [MARKS] until the app takes it ([takeMarks]) and
         * turns it into a real record — CardStore._takeWidgetMarks.
         */
        fun markDone(context: Context, cardId: String, today: LocalDate = LocalDate.now()) {
            val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            val day = today.toString()
            val edit = prefs.edit()
            prefs.getString(KEY, null)?.let { raw ->
                try {
                    val o = JSONObject(raw)
                    val cards = o.getJSONArray("cards")
                    for (i in 0 until cards.length()) {
                        val c = cards.getJSONObject(i)
                        if (c.optString("id") == cardId) {
                            val last = if (c.isNull("last")) null else c.optString("last")
                            if (last == null || last < day) c.put("last", day)
                        }
                    }
                    edit.putString(KEY, o.toString())
                } catch (_: Exception) {
                }
            }
            val marks = try {
                JSONArray(prefs.getString(MARKS, "[]"))
            } catch (_: Exception) {
                JSONArray()
            }
            marks.put(JSONObject().put("id", cardId).put("day", day))
            edit.putString(MARKS, marks.toString()).apply()
        }

        /** The marks not yet recorded by the app, as JSON; clears them. */
        fun takeMarks(context: Context): String {
            val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            val marks = prefs.getString(MARKS, null) ?: return "[]"
            prefs.edit().remove(MARKS).apply()
            return marks
        }

        /** `null` until the app has run once since the widgets shipped. */
        fun load(context: Context, today: LocalDate = LocalDate.now()): WidgetSnapshot? {
            val raw = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE).getString(KEY, null)
                ?: return null
            return try {
                parse(JSONObject(raw), today)
            } catch (_: Exception) {
                null
            }
        }

        private fun parse(o: JSONObject, today: LocalDate): WidgetSnapshot? {
            if (o.optInt("v") > VERSION) return null
            val t = o.getJSONObject("theme")
            val theme = ThemeColors(
                dark = t.optBoolean("dark"),
                surface = t.color("surface"),
                onSurface = t.color("onSurface"),
                muted = t.color("muted"),
                outline = t.color("outline"),
                primary = t.color("primary"),
                onPrimary = t.color("onPrimary"),
            )
            val strings = WidgetStrings.parse(o.optJSONObject("strings"))
            val tierJson = o.getJSONObject("tiers")
            val tiers = Tier.values().associateWith { tier ->
                val c = tierJson.getJSONObject(tier.key)
                TierColors(c.color("bg"), c.color("ink"), c.color("ring"), c.color("track"))
            }
            val list = o.getJSONArray("cards")
            val cards = (0 until list.length()).mapNotNull { i ->
                val c = list.getJSONObject(i)
                val id = c.optString("id")
                if (id.isEmpty()) return@mapNotNull null
                WidgetCard(
                    id = id,
                    name = c.optString("name"),
                    icon = c.optString("icon"),
                    last = if (c.isNull("last")) null else runCatching { LocalDate.parse(c.getString("last")) }.getOrNull(),
                    typical = if (c.isNull("typical")) null else c.optInt("typical").takeIf { it > 0 },
                    today = today,
                    strings = strings,
                )
            }
            return WidgetSnapshot(theme, o.optBoolean("doneButton", true), strings, tiers, cards)
        }

        private fun JSONObject.color(key: String): Int = getLong(key).toInt()
    }
}

/**
 * The text the widgets draw, as the app last published it
 * (`Strings.widgetStrings` in lib/l10n). `{n}` is the number this side works
 * out. The fallbacks are the Turkish originals, for a widget placed before
 * the app has ever run.
 */
internal class WidgetStrings(private val map: Map<String, String>) {
    private fun s(key: String, fallback: String) = map[key]?.takeIf { it.isNotEmpty() } ?: fallback
    private fun fmt(key: String, fallback: String, value: Int) = s(key, fallback).replace("{n}", value.toString())

    val title: String get() = s("title", "Kaç gün oldu?")
    fun late(n: Int): String = fmt("late", "{n} kart gecikti", n)
    val notMarked: String get() = s("notMarked", "Henüz işaretlenmedi")
    val learning: String get() = s("learning", "Ritim öğreniliyor")
    fun daysLeft(n: Int): String = fmt("daysLeft", "{n} gün kaldı", n)
    val dueToday: String get() = s("dueToday", "Bugün sırası")
    fun daysOver(n: Int): String = fmt("daysOver", "{n} gün geçti", n)
    val unitDays: String get() = s("unitDays", "gün oldu")
    val unitNone: String get() = s("unitNone", "kayıt yok")
    val doneToday: String get() = s("doneToday", "Bugün yapıldı")
    val markDone: String get() = s("markDone", "Bugün yaptım")
    val firstCard: String get() = s("firstCard", "İlk kartını oluştur")
    val openApp: String get() = s("openApp", "Kartlarını görmek için uygulamayı aç")

    companion object {
        val DEFAULT = WidgetStrings(emptyMap())

        fun parse(o: JSONObject?): WidgetStrings {
            if (o == null) return DEFAULT
            val map = HashMap<String, String>()
            for (key in o.keys()) o.optString(key).let { if (it.isNotEmpty()) map[key] = it }
            return WidgetStrings(map)
        }
    }
}

internal enum class Tier(val key: String) { FRESH("fresh"), CALM("calm"), SOON("soon"), LATE("late") }

internal class TierColors(val bg: Int, val ink: Int, val ring: Int, val track: Int)

internal class ThemeColors(
    val dark: Boolean,
    val surface: Int,
    val onSurface: Int,
    val muted: Int,
    val outline: Int,
    val primary: Int,
    val onPrimary: Int,
) {
    companion object {
        /** Kiremit, the default theme — for a widget placed before the app ever ran. */
        val DEFAULT = ThemeColors(
            dark = false,
            surface = 0xFFFBF5F0.toInt(),
            onSurface = 0xFF241F1B.toInt(),
            muted = 0xFF87786C.toInt(),
            outline = 0xFFE6D9CF.toInt(),
            primary = 0xFFB8492A.toInt(),
            onPrimary = 0xFFFFF3EA.toInt(),
        )
    }
}

private val FALLBACK_TIER = TierColors(
    0xFFFFFDFA.toInt(), 0xFF241F1B.toInt(), 0xFFC98A6B.toInt(), 0x1A241F1B,
)

/** One card, with everything the app's `statsFor` derives from it. */
internal class WidgetCard(
    val id: String,
    val name: String,
    val icon: String,
    val last: LocalDate?,
    val typical: Int?,
    today: LocalDate,
    strings: WidgetStrings = WidgetStrings.DEFAULT,
) {
    val days: Int = last?.let { ChronoUnit.DAYS.between(it, today).toInt().coerceAtLeast(0) } ?: 0

    val ratio: Double = if (typical != null && typical != 0) {
        days.toDouble() / typical
    } else {
        min(0.95, days / 30.0)
    }

    /** `+1` keeps short-interval cards from flagging on a single late day. */
    val isLate: Boolean = typical != null && last != null && days > typical * 1.25 + 1

    val tier: Tier = when {
        isLate -> Tier.LATE
        ratio > 0.95 -> Tier.SOON
        ratio < 0.5 -> Tier.FRESH
        else -> Tier.CALM
    }

    /** Ring fill, 3–100. */
    val pct: Int = (ratio * 100).roundToInt().coerceIn(3, 100)

    /** Days left in the usual interval; negative past it, `null` while learning. */
    val remaining: Int? = if (typical == null || last == null) null else typical - days

    /** The card's one status line — `CardStatus` in lib/widgets/card_status.dart. */
    val status: String = when {
        last == null -> strings.notMarked
        remaining == null -> strings.learning
        remaining > 0 -> strings.daysLeft(remaining)
        remaining == 0 -> strings.dueToday
        else -> strings.daysOver(-remaining)
    }

    /** Due today or past due: drawn in the accent, bold. */
    val urgent: Boolean = remaining != null && remaining <= 0

    /** Already recorded today: the "Bugün yaptım" button shows it's done. */
    val doneToday: Boolean = last == today

    /** The big number, or a dash when nothing is recorded yet. */
    val dayText: String = if (last == null) "—" else days.toString()
    val unitText: String = if (last == null) strings.unitNone else strings.unitDays
}
